File: gameshow-app/src/lib/wheelLogic.ts
/**
* Wheel Logic Management
* Handles spinning wheel mechanics, segments, and results
*/

import { WheelSegment, SpinResult, WheelConfig, GameAction } from '@/types/game';
import { WHEEL_CONFIG_8_SEGMENTS } from './wheelConfigurations';

// Default wheel segments configuration with 8 sections (using utility function)
const DEFAULT_SEGMENTS: Omit<WheelSegment, 'color'>[] = generateWheelSegments(WHEEL_CONFIG_8_SEGMENTS);

// Soft color palette for wheel segments
const SOFT_COLORS = [
	'#F8F9FA',  // Very light white
	'#F1F3F4',  // Light gray-white
	'#F5F5DC',  // Light beige
	'#F0E68C',  // Light khaki
	'#F0F8FF',  // Light azure
	'#F0FFF0',  // Light honeydew
	'#F5FFFA',  // Light mint cream
	'#FFF8DC',  // Light cornsilk
	'#FDF5E6',  // Light old lace
	'#F0F8FF',  // Light alice blue
	'#F5F5F5',  // Light gray
	'#FFE4E1',  // Light misty rose
	'#F0FFFF'   // Light azure
];

/**
 * Utility function to generate wheel segments with dynamic count and auto-calculated angles
 * @param segmentConfigs Array of segment configurations (text, action, id)
 * @param startAngle Starting angle for the first segment (default: 0)
 * @returns Array of wheel segments with calculated angles and random colors
 */
export function generateWheelSegments(
	segmentConfigs: Array<{ text: string; action: GameAction; id?: number }>,
	startAngle: number = 0
): Omit<WheelSegment, 'color'>[] {
	const segmentCount = segmentConfigs.length;
	const segmentAngle = 360 / segmentCount;
	
	return segmentConfigs.map((config, index) => ({
		id: config.id ?? index,
		text: config.text,
		action: config.action,
		angle: startAngle + (index * segmentAngle)
	}));
}

// Configuration constants
const DEFAULT_CONFIG: WheelConfig = {
	segments: [],
	canvasSize: 600,
	spinDuration: 3000,
	minSpins: 3,
	maxSpins: 8,
};

export class WheelManager {
	private config: WheelConfig;
	private isSpinning: boolean = false;

	constructor() {
		this.config = { ...DEFAULT_CONFIG };
		this.initializeSegments();
	}

	/**
	* Get random soft color
	*/
	private getRandomSoftColor(): string {
		return SOFT_COLORS[Math.floor(Math.random() * SOFT_COLORS.length)];
	}

	/**
	* Initialize wheel segments with randomized colors
	*/
	private initializeSegments(): void {
		this.config.segments = DEFAULT_SEGMENTS.map((segment, index) => ({
			...segment,
			color: this.getRandomSoftColor(),
			angle: (360 / DEFAULT_SEGMENTS.length) * index
		}));
	}

	/**
	* Get wheel configuration
	*/
	getConfig(): WheelConfig {
		return { ...this.config };
	}

	/**
	* Get all wheel segments
	*/
	getSegments(): WheelSegment[] {
		return [...this.config.segments];
	}

	/**
	* Randomize segment colors
	*/
	randomizeColors(): void {
		this.config.segments.forEach(segment => {
			segment.color = this.getRandomSoftColor();
		});
	}

	/**
	* Randomize segment order
	*/
	randomizeSegments(): void {
		// Shuffle segments using Fisher-Yates algorithm
		const segments = [...this.config.segments];
		for (let i = segments.length - 1; i > 0; i--) {
			const j = Math.floor(Math.random() * (i + 1));
			[segments[i], segments[j]] = [segments[j], segments[i]];
		}

		// Recalculate angles for shuffled segments
		const segmentAngle = 360 / segments.length;
		segments.forEach((segment, index) => {
			segment.angle = index * segmentAngle;
			segment.color = this.getRandomSoftColor();
		});

		this.config.segments = segments;
	}

	/**
	* Generate spin parameters for animation
	*/
	generateSpin(): { totalRotation: number; duration: number } {
		// Calculate spin parameters
		const baseRotations = 5; // Always spin about 5 full rotations
		const extraRotations = Math.random() * 2; // Add 0-2 extra rotations
		const totalRotations = (baseRotations + extraRotations) * 360;

		// Generate a random final angle (where the wheel stops)
		const randomFinalAngle = Math.random() * 360;
		const totalRotation = totalRotations + randomFinalAngle;

		return {
			totalRotation,
			duration: this.config.spinDuration
		};
	}

	/**
	* Calculate winner based on final wheel position and optionally snap to center
	*/
	calculateWinnerAtPosition(finalRotation: number, snapToCenter: boolean = true): SpinResult {
		const segmentAngle = 360 / this.config.segments.length;
		
		// Normalize the rotation
		const normalizedRotation = ((finalRotation % 360) + 360) % 360;
		
		// Calculate the effective rotation for determining the winning segment
		// The pointer is at 90 degrees (3 o'clock position)
		const effectiveRotation = (360 - normalizedRotation) % 360;
		const segmentIndex = Math.floor(effectiveRotation / segmentAngle) % this.config.segments.length;
		const selectedSegment = this.config.segments[segmentIndex];

		let adjustedRotation = finalRotation;
		
		if (snapToCenter) {
			// Calculate the center angle of the winning segment
			const segmentCenterAngle = segmentIndex * segmentAngle + (segmentAngle / 2);
			
			// Calculate how much we need to adjust to center the segment at the pointer
			const targetRotation = (360 - segmentCenterAngle) % 360;
			const currentEffectiveRotation = effectiveRotation % 360;
			
			// Find the smallest adjustment needed
			let adjustment = (targetRotation - currentEffectiveRotation + 360) % 360;
			if (adjustment > 180) adjustment -= 360;
			
			adjustedRotation = finalRotation + adjustment;
		}

		const result: SpinResult = {
			segment: selectedSegment,
			finalAngle: adjustedRotation,
			duration: this.config.spinDuration,
			winnerText: selectedSegment.text.toUpperCase()
		};

		return result;
	}

	/**
	* Start spinning - sets spinning state
	*/
	startSpin(): void {
		this.isSpinning = true;
	}

	/**
	* Stop spinning - clears spinning state
	*/
	stopSpin(): void {
		this.isSpinning = false;
	}

	/**
	* Legacy spin method for compatibility - now just generates spin parameters
	*/
	generateSpinParams(): { totalRotation: number; duration: number } {
		return this.generateSpin();
	}

	/**
	* Check if wheel is currently spinning
	*/
	getIsSpinning(): boolean {
		return this.isSpinning;
	}

	/**
	* Update spin duration
	*/
	setSpinDuration(duration: number): void {
		this.config.spinDuration = Math.max(1000, Math.min(10000, duration));
	}

	/**
	* Get wheel statistics
	*/
	getStats() {
		const actionCounts = this.config.segments.reduce((acc, segment) => {
			acc[segment.action] = (acc[segment.action] || 0) + 1;
			return acc;
		}, {} as Record<string, number>);

		return {
			totalSegments: this.config.segments.length,
			actions: Object.keys(actionCounts),
			actionCounts,
			spinDuration: this.config.spinDuration,
			isSpinning: this.isSpinning
		};
	}

	/**
	* Reset to default configuration
	*/
	resetToDefaults(): void {
		this.config = { ...DEFAULT_CONFIG };
		this.initializeSegments();
		this.isSpinning = false;
	}

	/**
	* Set wheel configuration with any number of segments
	* Automatically calculates angles based on the number of segments provided
	* @param segmentConfigs Array of segment configurations (text, action, optional id)
	* @param startAngle Optional starting angle for the first segment (default: 0)
	*/
	setWheelSegments(
		segmentConfigs: Array<{ text: string; action: GameAction; id?: number }>,
		startAngle: number = 0
	): void {
		const segments = generateWheelSegments(segmentConfigs, startAngle);
		this.config.segments = segments.map((segment) => ({
			...segment,
			color: this.getRandomSoftColor()
		}));
	}

	/**
	* Calculate which segment the pointer is pointing to based on rotation
	*/
	getSegmentAtAngle(rotation: number): WheelSegment | null {
		const normalizedRotation = ((rotation % 360) + 360) % 360;
		const segmentAngle = 360 / this.config.segments.length;

		// The pointer points to the right (3 o'clock position = 90 degrees)
		// We need to find which segment is at that position
		const effectiveRotation = (360 - normalizedRotation) % 360;
		const segmentIndex = Math.floor(effectiveRotation / segmentAngle) % this.config.segments.length;

		return this.config.segments[segmentIndex] || null;
	}
}

// Export singleton instance
export const wheelManager = new WheelManager();

File: gameshow-V2/server/wheel_logic.py
"""
Wheel Logic Management
Handles spinning wheel mechanics, segments, and results
"""

import random
import math
from typing import Dict, List, Tuple, Optional

# Default wheel segments configuration
DEFAULT_SEGMENTS = [
    {"id": 0, "text": "New Rule", "action": "new_rule", "angle": 0},
    {"id": 1, "text": "New Rule", "action": "new_rule", "angle": 30},
    {"id": 2, "text": "New Rule", "action": "new_rule", "angle": 60},
    {"id": 3, "text": "Modify: Audience Choice", "action": "audience_choice", "angle": 90},
    {"id": 4, "text": "Modify: Audience Choice", "action": "audience_choice", "angle": 120},
    {"id": 5, "text": "Challenge", "action": "challenge", "angle": 150},
    {"id": 6, "text": "Challenge", "action": "challenge", "angle": 180},
    {"id": 7, "text": "Challenge", "action": "challenge", "angle": 210},
    {"id": 8, "text": "Modify: Duplicate", "action": "duplicate", "angle": 240},
    {"id": 9, "text": "Modify: Reverse", "action": "reverse", "angle": 270},
    {"id": 10, "text": "Modify: Swap", "action": "swap", "angle": 300}
]

# Configuration constants
MAX_SEGMENTS = 12
DEFAULT_SPIN_DURATION = 3000
DEFAULT_MIN_SPINS = 3
ANGLE_RANDOMNESS = 15

class WheelLogic:
    def __init__(self):
        # Game show wheel segments with special actions - no styling concerns
        self.segments = DEFAULT_SEGMENTS.copy()
        
        self.current_segment = None
        self.is_spinning = False
        self.spin_duration = DEFAULT_SPIN_DURATION
        self.min_spins = DEFAULT_MIN_SPINS
        
        # Initialize angles properly
        self._recalculate_angles()
        
    def get_segments(self) -> List[Dict]:
        """Get all wheel segments"""
        return self.segments.copy()
    
    def get_current_segment(self) -> Optional[Dict]:
        """Get the currently selected segment"""
        return self.current_segment
    
    def add_segment(self, text: str, action: str) -> bool:
        """Add a new segment to the wheel"""
        if len(self.segments) >= MAX_SEGMENTS:
            return False
        
        new_id = max(segment["id"] for segment in self.segments) + 1
        angle = (360 / (len(self.segments) + 1)) * len(self.segments)
        
        new_segment = {
            "id": new_id,
            "text": text,
            "action": action,
            "angle": angle
        }
        
        self.segments.append(new_segment)
        self._recalculate_angles()
        return True
    
    def remove_segment(self, segment_id: int) -> bool:
        """Remove a segment from the wheel"""
        for i, segment in enumerate(self.segments):
            if segment["id"] == segment_id:
                del self.segments[i]
                self._recalculate_angles()
                return True
        return False
    
    def spin(self) -> Dict:
        """Spin the wheel and return the result"""
        if self.is_spinning:
            return {"error": "Wheel is already spinning"}
        
        self.is_spinning = True
        
        try:
            # Select random segment
            selected_segment = random.choice(self.segments)
            self.current_segment = selected_segment
            
            # Calculate final rotation angle
            final_angle = self._calculate_final_angle(selected_segment)
            
            # Simulate spin result
            result = {
                "segment": selected_segment,
                "final_angle": final_angle,
                "duration": self.spin_duration,
                "min_spins": self.min_spins
            }
            
            return result
        finally:
            self.is_spinning = False
    
    def _calculate_final_angle(self, segment: Dict) -> float:
        """Calculate the final rotation angle for a segment"""
        # Add some randomness to make it feel more natural
        base_angle = segment["angle"]
        random_offset = random.uniform(-ANGLE_RANDOMNESS, ANGLE_RANDOMNESS)
        final_angle = base_angle + random_offset
        
        # Ensure angle is between 0 and 360
        final_angle = final_angle % 360
        return final_angle
    
    def _recalculate_angles(self) -> None:
        """Recalculate angles for all segments after adding/removing"""
        segment_count = len(self.segments)
        if segment_count == 0:
            return
        
        angle_step = 360 / segment_count
        for i, segment in enumerate(self.segments):
            segment["angle"] = i * angle_step
    
    def set_spin_duration(self, duration: int) -> None:
        """Set the spin duration in milliseconds"""
        self.spin_duration = max(1000, min(10000, duration))  # Between 1-10 seconds
    
    def set_min_spins(self, spins: int) -> None:
        """Set the minimum number of full rotations"""
        self.min_spins = max(1, min(10, spins))  # Between 1-10 spins
    
    def get_wheel_stats(self) -> Dict:
        """Get wheel statistics"""
        return {
            "total_segments": len(self.segments),
            "actions": list(set(segment["action"] for segment in self.segments)),
            "spin_duration": self.spin_duration,
            "min_spins": self.min_spins
        }
    
    def reset_to_defaults(self) -> None:
        """Reset wheel to default segments"""
        self.segments = DEFAULT_SEGMENTS.copy()
        self._recalculate_angles()
        self.current_segment = None
